### PR TITLE
Introduction to CI exercise

### LINKED ISSUE #32: Introduction to CI
## Brief
Even simple changes to a code base can cause serious bugs. As the number of lines of code go up so do the chances of making a mistake. Chuck in multiple team members all making changes to the same code base and the chances of broken code making its way to production becomes almost a certainty.

> Continuous Integration (CI) is a development practice that requires developers to integrate code into a shared repository several times a day. Each check-in is then verified by an automated build, allowing teams to detect problems early. 
> By integrating regularly, you can detect errors quickly, and locate them more easily. [ThoughtWorks](https://www.thoughtworks.com/continuous-integration)

CI is a vital discipline that we want to be able to teach to anyone whom intends to release code.

This is the first exercise in a Module on CI. The purpose of this exercise to introduce the topic and get participants to see and fix a failing build.

## Learning Objectives
- Introduce topic of CI
- Introduce CI Server and its outputs

## Suggested Exercise
Participants could start a pre-canned CI Server which is already running a failing set of tests. The Participant could fix the failing test and see that reflected in the CI server.

## Acceptance Criteria:
The readme to the exercise should include the following:

### Introduction
`# Required - Give a written explanation of the purpose of the topic this exercise is based on`

## Required prerequisite knowledge
Having some experience of testing is useful but should not be essential.

## Useful Terminology
`# Optional - any key terms you intend to use in your exercise, if appropriate, could be listed here`

### Tutorial
`# Ideal but optional - This section should provided a guided exercise showing participants how to use the subject of the exercise`

### Exercise
`# required - Following the tutorial, provide an exercise that requires the participant use the knowledge that they have gained through the intro and tutorial`

### Verification test
This will be seeing the build go green having fixed and comited the failing test.

## Other
`# Optional - add any other requirements here`

### PR DESCRIPTION
**Pull request by lashd**
First draft of the Introduction to CI exercise. 

This will resolve #32 

In this pull request:
- Exercise
- Support for after hook in cic up command
- various tidyups (see commits)

### CODE REVIEW

File: exercises/CI/introduction_to_ci/README.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+Continuous integration servers are essentially glorified schedulers. They sit looking at your Version Control System (VCS) and when they see that a change has been introduced they do what ever they have been configured to do. E.g. to compile the new version of the code base and run a set of tests to see if it still works. Different CI servers may have different user interfaces, however they essentially all do the same thing and have a very common feature set. Therefore learn how to use one CI server and you should find that switching to another is a pretty easy process.
+
+## Exercise Learning Objectives
+- Introduction Concourse CI
+
+## Exercise
+**Note:** Before going any further do the following:
+- `cd YOUR_CLONE_OF_THIS REPO`
+- `source ./bin/env`
+- `cd `
```
Comment (jimconner) [inline exercises/CI/introduction_to_ci/README.md]: The cd command here on line 26 seems to have lost it's exerise path.

File: exercises/CI/introduction_to_ci/README.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+In this exercise you will be working using a CI server called Concourse. We've done the hard bit and have provided you with your very own Concourse installation. We've also provided you with a GIT server for your CI server to talk to. To start everything up, run: cic up
```
Comment (jimconner) [inline exercises/CI/introduction_to_ci/README.md]: The `cic up` at the end of this line is not appearing as a command (missing backticks?)

File: exercises/CI/introduction_to_ci/README.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+- `source ./bin/env`
+- `cd `
+
+In this exercise you will be working using a CI server called Concourse. We've done the hard bit and have provided you with your very own Concourse installation. We've also provided you with a GIT server for your CI server to talk to. To start everything up, run: cic up 
+
+If all is well you should see the following:
+```
+  [OK] Environment cic'd up :)
+
+```
```
Comment (jimconner) [inline exercises/CI/introduction_to_ci/README.md]: Apparently even if all is not well, you'll see the same [OK] Environment cic'd up :) message

```
8055dd6791ed: Pull complete 
b749918e3aa2: Pull complete 
75c387007b66: Pull complete 
2b88a7cdad8f: Pull complete 
Digest: sha256:d6479f769d18b2d7d2f654a46d7715bd6a1ec5171870c7c8c58cdb1525a65815
Status: Downloaded newer image for lvlup/ci_course:0.5.3


Identity added: git-server/keys/id_rsa (git-server/keys/id_rsa)
Cloning into 'checkout'...
Warning: Permanently added 'git-server,172.18.0.3' (ECDSA) to the list of known hosts.
error: refs/heads/master does not point to a valid object!
fatal: git upload-pack: not our ref 5fc17da4781cb3df5f3c670b2007c10646a0f415
fatal: The remote end hung up unexpectedly
[ERROR] Failed to cic up the environment see above output for details

[OK] Environment cic'd up :)
```

File: exercises/CI/introduction_to_ci/README.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+Click the login button at the top right of the [concourse home page](http://127.0.0.1:8080) and login with username `test` and password `test`. You'll note that there arn't currently any 'pipelines' configured. A pipeline is the name that Concourse gives to one or more tasks/jobs that have been configured to be run in a prescribed order when a particular event happens. E.g. a change is detected in GIT. Pipelines can be can be as simple as a one step process or they can be used to model very complex processes. In a later exercise we'll go in to pipelines and there features in more depth.
+
+We have supplied a very simple pipeline. Let's put it in to concourse.
+
+Some CI's have extensive UIs through which they can be configured, Concourse is primarly configured via it's commandline interface (CLI) which makes webservice calls to configure the Concourse server. The CLIs for each O/S are available in `./resources` directory. Use the fly binary that is correct for your O/S to complete this exercise.
+
+First we must login, run `./resources/linux/fly -t local login --concourse-url 127.0.0.1:8080 -u test -p test`
```
Comment (jimconner) [inline exercises/CI/introduction_to_ci/README.md]: Need http:// at the start of the URL here.

```
jim@Minty ~/development/ci-cd-training/exercises/CI/introduction_to_ci (32_introduction_to_ci) $ ./resources/linux/fly -t local login --concourse-url 127.0.0.1:8080 -u test -p test
logging in to team 'main'

could not reach the Concourse server called local:

    parse 127.0.0.1:8080: first path segment in URL cannot contain colon

is the targeted Concourse running? better go catch it lol
```

File: exercises/CI/introduction_to_ci/README.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+
+
+Visit the [concourse home page](http://127.0.0.1:8080) again and you'll see the pipeline that the was just pushed.
+![Pending Pipeline](./pending_pipeline_screenshot.png)
+
+
+You'll notice the following:
+- there is now a pipe line called: `ci-intro-pipeline`
+- the pipeline is paused
```
Comment (jimconner) [inline exercises/CI/introduction_to_ci/README.md]: How will they notice that it is paused? We need to tell them that a blue bar at the top of a concourse job means it is paused.

File: exercises/CI/introduction_to_ci/README.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+When pipelines are first pushed, they are not enabled. In order to run the pipeline we need to unpause it. Do this by either clicking the play symbol or by running: `./resources/linux/fly -t local unpause-pipeline -p ci-intro-pipeline`. Unpausing the pipeline should cause Concourse to trigger a build for our new pipeline. Go back to the [detailed view of the pipeline](127.0.0.1:8080/teams/main/pipelines/ci-intro-pipeline) after 20 seconds or so, you'll notice that that pipeline goes green. This means that the code has been pulled successfully from GIT and that the singe job, that is in the pipeline, has executed successfully.
```
Comment (jimconner) [inline exercises/CI/introduction_to_ci/README.md]: Play symbol is not visible from the screen we took them to... Need to click on the fan icon in the top left to get back to the pipelines view where the play button is shown.

File: exercises/CI/introduction_to_ci/README.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+First we must login, run `./resources/linux/fly -t local login --concourse-url 127.0.0.1:8080 -u test -p test`
+you should see the following output:
+```
+logging in to team 'main'
+
+
+target saved
+```
+
+Now let's push a pipeline to concourse by running: `./resources/linux/fly  -t local set-pipeline -p ci-intro-pipeline -c ./resources/pipeline.yml -n`
```
Comment (jimconner) [inline exercises/CI/introduction_to_ci/README.md]: Before we push the pipeline, we probably need a bit where the user takes a look at the pipeline file, and we explain a little about what this pipeline does, and how concourse uses 'resources' and 'tasks' to build it's pipelines automatically.

File: exercises/CI/introduction_to_ci/README.md [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+![Pending Pipeline](./pending_pipeline_screenshot.png)
+
+
+You'll notice the following:
+- there is now a pipe line called: `ci-intro-pipeline`
+- the pipeline is paused
+
+When pipelines are first pushed, they are not enabled. In order to run the pipeline we need to unpause it. Do this by either clicking the play symbol or by running: `./resources/linux/fly -t local unpause-pipeline -p ci-intro-pipeline`. Unpausing the pipeline should cause Concourse to trigger a build for our new pipeline. Go back to the [detailed view of the pipeline](127.0.0.1:8080/teams/main/pipelines/ci-intro-pipeline) after 20 seconds or so, you'll notice that that pipeline goes green. This means that the code has been pulled successfully from GIT and that the singe job, that is in the pipeline, has executed successfully.
+
+![Passing Pipeline](./passing_pipeline_screenshot.png)
```
Comment (jimconner) [inline exercises/CI/introduction_to_ci/README.md]: In my case, concourse is unable to connect to the git server, so pipeline never goes green for me.

**Comment by jimconner** [comment]
Many inline comments left in the README.md as I attempted to work through this exercise. git serer did not work properly for me, so concourse was unable to run through the pipeline at all.
I feel that the current pipeline file is too simplistic for this exercise and that we should enhance it to something that runs some kind of actual test rather than just printing 'Hello World' in a shell script.

File: exercises/CI/introduction_to_ci/README.md [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+### Now it's your turn
+You are part of a team who has been working on a new application, a team member has carefully made a change to the code and is recommending that the change be released straight to production. Nothing could possibly go wrong!... could it?
+
+to complete this exercise do the following:
+
+1. Push the project pipeline to concourse - The pipeline file can be found in `./resources/application-pipeline.yml`
+**Note:** Make sure that it is enabled and runs successfully.
+2. The proposed change sits uncommitted in `./resources/checkout`. Commit and push your team mate's changes to GIT
```
Comment (jimconner) [inline exercises/CI/introduction_to_ci/README.md]: I cannot find the resources/checkout folder/file here. 
Also , it is not clear to me here what git magic was applied to make this work. We haven't mentioned in the readme that we're even spinning up a git instance for the purposes of this exervise, nor that we've pre-seeded it and pre-configured the command line to use it.

Comment (lashd) [reply exercises/CI/introduction_to_ci/README.md]: This happened because the checkout in the after script failed. It'll be there when we get that fixed

**Comment by lashd** [comment]
We do actually already tell the reader what we are spinning up for them when they run `cic up` [here](https://github.com/lvl-up/ci-cd-training/blame/32_introduction_to_ci/exercises/CI/introduction_to_ci/README.md#L28)

I've tried to avoided the port number that git is running on in this section as I want to avoid the readme becoming brittle should we change it.

I'll show the reader that the checkout has indeed come from a server running on their machine.

**Pull request merged**